### PR TITLE
tsdb: Eliminate the goroutine leak calling tests.

### PR DESCRIPTION
**Pull request by johncming**
A tool imported into repo to find possible goroutine leak https://github.com/prometheus/prometheus/pull/7566.
The deprecated `wal` implementations have a situation that causes goroutine leak. It leads to test failure.

The leak bug result like
```
=== Failed
=== FAIL: tsdb  (0.00s)
PASS
goleak: Errors on successful test run: found unexpected goroutines:
[Goroutine 15483 in state chan send, with github.com/prometheus/prometheus/tsdb.(*SegmentWAL).cut.func2 on top of the stack:
goroutine 15483 [chan send]:
github.com/prometheus/prometheus/tsdb.(*SegmentWAL).cut.func2(0xc0000be210)
	/home/circleci/project/tsdb/wal.go:599 +0xc0
created by github.com/prometheus/prometheus/tsdb.(*SegmentWAL).cut
	/home/circleci/project/tsdb/wal.go:598 +0x160
]
FAIL	github.com/prometheus/prometheus/tsdb	174.747s


DONE 1022 tests, 17 skipped, 1 failure in 379.805s
make: *** [Makefile.common:172: common-test] Error 1

Exited with code exit status 2
```

The PR has relation with another PR https://github.com/prometheus/prometheus/pull/7655#issuecomment-666016301

### CODE REVIEW

File: tsdb/wal.go [hunk 1]
```diff
@@ -673,7 +673,14 @@ func (w *SegmentWAL) run(interval time.Duration) {
 		defer ticker.Stop()
 		tick = ticker.C
 	}
-	defer close(w.donec)
+	defer func() {
+		// Drain and process any remaining functions.
+		for f := range w.actorc {
+			//nolint: errcheck
+			f()
```
Comment (codesome) [inline tsdb/wal.go:680]: Let us log this error like below.

**Review by codesome — APPROVED** [review]
LGTM with 1 comment

File: tsdb/wal.go (hunk 1 above)
Comment (johncming) [reply tsdb/wal.go:680]: OK. I will add it.

**Review by codesome — CHANGES_REQUESTED** [review]
Sorry, I take my LGTM back. If it was possible to add a function to the actor after stopping (even though we are draining it already before exiting), then it is possible that function will be added to the actor even after `close(w.actorc)`, hence cause a panic. We should actually reject WAL Logging after closing, but this doesn't eliminate the mentioned race as a WAL Logging could be on-going.

I don't have a way to fix this on top of my mind.

**Comment by johncming** [comment]
> Sorry, I take my LGTM back. If it was possible to add a function to the actor after stopping (even though we are draining it already before exiting), then it is possible that function will be added to the actor even after `close(w.actorc)`, hence cause a panic. We should actually reject WAL Logging after closing, but this doesn't eliminate the mentioned race as a WAL Logging could be on-going.
> 
> I don't have a way to fix this on top of my mind.

Sorry. This PR can't solve the leak.  And close the channel in receiver goroutine will lead to panic. 

Since it has been marked as deprecated, I try to eliminate leak with minimal changes.

**Comment by roidelapluie** [comment]
cc @codesome can we close this pull request? Thanks!

**Comment by bwplotka** [comment]
Is this still reproducible? Let's reopen and start new PR if this happens again :hugs:

**Pull request closed without merging**